Preview SVG files

Chris2011 opened the pull request:
The PR will preview SVG files inside the IDE with MultiViewSupport

![image](https://github.com/user-attachments/assets/c7367c94-4be3-4f06-b09b-976934f46d4e)

I already add actions that make sense to the SVGDataObject as for XML files like:
* View (Opens in browser)
* Check XML and Validate XML
* Reload Document (Disabled atm unfortunately)

I have 2 problems faced here:
1. I can't find an implementation for Reload Document. Adding the action to mf-layers.xml file isn't enough without having the cookie set in the class as you can see for View, CheckXML and ValidateXML.
2. I have problems showing embedded images with relative paths:
```xml
<svg xmlns="http://www.w3.org/2000/svg" width="300" height="200">
    <image href="icons/vue.png" x="10" y="10" height="180px" width="280px"/>
</svg>
```

For problem 1. reload document for XML files were implemented, when there was an assumption that some references can change and the IDE can't recognize them. I really don't know whether I need this action or still this action is relevant for SVG. If you have a hint please lemme know.

For problem 2. I didn't want to add custom logic but I need to think about to handle this. Also I don't know whether batik can handle external resources as relativ paths. I have another image with an base64 encoded image and this is working just fine, but not to load from the path. VS Code and the browser can handle it quite well.

I got the image from https://commons.wikimedia.org/wiki/File:SVG_Logo.svg it is CC-BY-SA-4.0 I used it, converted it to PNG and will now use Apache, is this correct?

Otherwise CC-BY-SA-4.0 is missing in nbbuild/licenses folder. Does that means, that Apache 2.0 is not compatible with CC-BA-SA-4.0 or just not up to date?

mbien commented on platform/libs.batik.read/release/modules/ext/batik-anim-1.18.jar:
curious if the paperwork job catches this since you committed two jars by accident

Chris2011 replied:
Thx, I see. I dunno why this happened. I thought this was just due to make them public to use in another ant module but it seems not. I will change this and remove this.

mbien reviewed (commented):
one thing to watch out for is that we might switch from batik to JSVG https://github.com/apache/netbeans/pull/7463#issuecomment-2252299955 at some point

not sure if anyone has this on the radar atm but it would be good to keep batik specific usages low.

Chris2011 commented:
> one thing to watch out for is that we might switch from batik to JSVG [#7463 (comment)](https://github.com/apache/netbeans/pull/7463#issuecomment-2252299955) at some point
> 
> not sure if anyone has this on the radar atm but it would be good to keep batik specific usages low.

Thx for the hint, I didn't noticed that. When this happens, of course I'm willing to switch to this. For now I would stick with batik if this is ok.

neilcsmith-net commented:
> > one thing to watch out for is that we might switch from batik to JSVG [#7463 (comment)](https://github.com/apache/netbeans/pull/7463#issuecomment-2252299955) at some point
> > not sure if anyone has this on the radar atm but it would be good to keep batik specific usages low.
> 
> Thx for the hint, I didn't noticed that. When this happens, of course I'm willing to switch to this. For now I would stick with batik if this is ok.

IMO, the two things don't need to be related.  Possibly switching to JSVG (or build time generation) for icons in the core platform doesn't necessarily mean we don't keep Batik around as an autoload library elsewhere in the IDE for other purposes.  In line with that, I would look at adding the additional Batik libraries required for this separately in the ide cluster rather than platform cluster.  The size of Batik already caused enough of a concern in #1278 - let's not add to that.

As and when anyone does make the icon loading use something else, all of the Batik libraries (along with a couple of dependencies that were previously moved), could move to the ide cluster?

I like the general idea of providing this support in the IDE incidentally.

cc/ @eirikbakke

Chris2011 commented:
@neilcsmith-net thx for the info. So as I understand it correctly, I should add those required libs to the XML module where I added the SVGDataObject, instead of the batik ant module, right?

neilcsmith-net commented:
I'd consider adding a `libs.batik` module in the ide cluster that relies on `libs.batik.read` in the platform cluster.  If we ever switch to something else for icons, all the external libraries could end up in the former.  But let's see what @mbien or @eirikbakke or anyone else think on that approach before you put time into it!

eirikbakke commented:
Thanks for this PR!

> Possibly switching to JSVG (or build time generation) for icons in the core platform doesn't necessarily mean we don't keep Batik around as an autoload library elsewhere in the IDE for other purposes.

I agree with this. I think it's OK to rely on Batik for IDE features that actually relate to SVG files, like this one. Batik probably supports a larger part of the SVG spec than the lighter-weight alternatives.

Whereas for NetBeans icons, it could make sense to switch to a lighter library, which just needs to work for the particular SVG files are used in NetBeans.

mbien commented:
IMO: if batik is supposed to stay we shouldn't switch to JSVG. The main motivation there would be to reduce dependencies (and potentially also become more resource efficient due to the changed impl). If batik is already there anyway, adding another lib just for icons is making it worse not better.

In other words: replacing batik with jsvg would be something worth exploring IMO, maintaining a lib wrapper for jsvg _additionally_ to batik should only be done if it is also demonstrated that there is a noteworthy memory and/or performance benefit in practice.

In reality you probably want to see how svgs look in a browser not in swing, having preview for it is more of a convenience thing. If it wouldn't support some SVG features it might be still ok? NB will never be inkscape ;)

eirikbakke commented:
@mbien Good point, two libraries is kind of awkward...

> In reality you probably want to see how svgs look in a browser not in swing, having preview for it is more of a convenience thing. If it wouldn't support some SVG features it might be still ok? NB will never be inkscape ;)

Probably true. (I don't know exactly what the limitations with jSVG are compared to Batik.)

troizet commented:
> In reality you probably want to see how svgs look in a browser not in swing, having preview for it is more of a convenience thing. If it wouldn't support some SVG features it might be still ok? NB will never be inkscape ;)

It would be really nice if there was an option to preview svg in the navigator, similar to other images:
![image_navigator_preview](https://github.com/user-attachments/assets/086cde84-47be-4623-a306-b986a6113053)

Chris2011 commented:
@troizet good catch, will check this too.

neilcsmith-net commented:
> If batik is already there anyway, adding another lib just for icons is making it worse not better.

I think these are mostly unrelated concerns, and probably don't need to be discussed on this PR ...

> maintaining a lib wrapper for jsvg additionally to batik should only be done if it is also demonstrated that there is a noteworthy memory and/or performance benefit in practice.

This is the only reason for switching for icons _at all_ IMO.  JSVG claims a ~98% less memory usage than Batik.  We're not sure if this is true in practice or whether we have an issue as yet.  If as more SVG icons are added we start to see startup time or memory concerns, then JSVG _might_ be a way forward.  As might be compile-time generation of bitmaps and/or Java classes.  FlatLaf has icons that are Java coded mirrors of SVG, and use of [Radiance's converter](https://github.com/kirill-grouchnikov/radiance/blob/sunshine/docs/tools/svg-transcoder/svg-transcoder.md) could also be a consideration.

Libraries to enable SVG previewing only need to be loaded if the user edits such a file.  Therefore there are different considerations.  Accuracy might trump overheads.  _If_ we did switch to JSVG for icons, and _if_ we felt JSVG was good enough for previewing, that would be the time to consider whether we need one library or two.  The key thing here, for me, is whether we want the feature?

At the moment, if we need to add additional Batik libraries purely for previewing SVG files, then I'd like to see them added separately (ie. in the IDE cluster) to the base library support for icons so we at least don't increase the weight of things in the base platform cluster.

neilcsmith-net commented:
> Having the option to trade ~17 jars against one sounds like a good trade all by itself (esp in context of how dependencies are managed in the NB build) 

Sure, I made that point when it was introduced in the first place (and pretty much the same as above) - https://github.com/apache/netbeans/pull/1278#issuecomment-500228647 and https://github.com/apache/netbeans/pull/1278#issuecomment-500232659

If number of jars is the primary concern here, let's just ship `batik-all.jar`!

eirikbakke commented:
I implemented the changes needed to switch ImageUtilities from Batik to JSVG in https://github.com/apache/netbeans/pull/7941 . I'll need to test it for a while to see that all icons appear correctly etc., including in my own NetBeans Platform app. (Out of time for today...)

Chris2011 commented:
Yes, I thought about adding scaling stuff and some other stuff into it and I also want to implement this, but not for this MVP. But at the end, we can leave this open until I:
* Implemented preview for the navigator panel
* Implementing the other stuff for scaling
* Move this stuff to SVG or an own package

Thx for the hints all.

matthiasblaesing commented:
This https://www.apache.org/legal/resolved.html#cc-sa indicates to me, that we should keep away from "CC-SA" licensed media. This might be an alternative (CC-0 licensed): https://www.svgrepo.com/svg/255830/svg. Alternatively, maybe @eirikbakke could be persuaded to draw an icon?

eirikbakke commented:
@matthiasblaesing Sure, here's an icon you're free to use:

![fileSVG](https://github.com/user-attachments/assets/02a2f8d8-cf9a-47a4-a405-70f731cb1fa0)

It's in the style of other NetBeans file type icons, which are generally of the form "tiny art inside a piece of paper with a folded corner". The "art" was drawn freehand as a hint to the SVG logo.

<img width="315" alt="image" src="https://github.com/user-attachments/assets/bee18085-e7c8-445e-90e0-95292eb23c69">

matthiasblaesing commented:
@eirikbakke thank you!
@Chris2011 thats one problem solved ;-)

Chris2011 commented:
I dunno whether everyone knows that little indication so isn't it possible to put the whole SVG icon inside this "file" but that's not that important, I'm fine with everything which doesn't look like the XML icon :). @eirikbakke thx for the suggestion.

eirikbakke commented:
Here's an alternative that alludes to a a line or bezier curve being edited.

![fileSVG](https://github.com/user-attachments/assets/e057f9da-e9fb-43f7-8c2d-b7790b935ce6)

eirikbakke commented:
> I have problems showing embedded images with relative paths:

I would probably prefer just disallowing loading of external resources, and require that any graphics be self-contained within the SVG file. An SVG file that tries to load external resources is going to break in a lot of contexts anyway, and it's useful for the preview to reflect this fact by not drawing that part of the image.

For example, when ImageUtilities loads SVG files, it explicitly disallows loading of external resources (both in the old Batik version and the new JSVG version at https://github.com/apache/netbeans/pull/7941 ).

Chris2011 commented:
So as suggested moving SVG to an own module or to image or wherever it fits best. I can think about image.svg ant module also with this name or directly into the ide/image module. Honestly, I don't care much for me it should be simple. Do you have any preferences? Also with the pros and cons for your suggestions. That would be nice.

matthiasblaesing commented:
Integration into `ide/image`:
  - Pro: GUI with scaling exists and Navigator with preview is implemented
  - Contra: Requires additional dependency on JSVG and xml module

Integration into `ide/xml`:
  - Pro: xml modules is not required
  - Contra: Requires additional dependency on JSVG
  - Contra: Image viewer and navigator need to be copied from `ide/image` or to be reimplemented
  - Contra: What makes SVG special compared to other XML based formats?

New Module:
  - Pro: Clean slate and enough space to implement new functions
  - Contra: Image viewer and navigator need to be copied from `ide/image` or to be reimplemented

Stuffing the new functions into existing modules seems like the easier road, but a new module seems to be correcter road.

eirikbakke commented:
I have merged https://github.com/apache/netbeans/pull/7941 now, which switches ImageUtilities from Batik to the lighter JSVG library. Eventually we will remove Batik, per https://github.com/apache/netbeans/issues/7969, but for now the libs.batik.read library remains, which means this PR should rebase cleanly on top of the latest master.

With respect to the SVG library switch, I'd suggest now (1) rebasing, (2) adding a dependency on the "libs.jsvg" dependency, (3) switching the Batik-related code to use the JSVG APIs instead, (4) verifying that it works, and then (5) removing the dependency on "libs.batik.read" and the changes relating to updating the Batik libraries. (Sorry for the extra work of switching SVG libraries!)

Chris2011 commented:
> Integration into `ide/image`:
> 
> * Pro: GUI with scaling exists and Navigator with preview is implemented
> * Contra: Requires additional dependency on JSVG and xml module
> ...

Thx @matthiasblaesing I already created a new module called image.svg. I also can rename it just to SVG.

Just curious, I had a look into the ImageViewer that you already mentioned, of course, when I reuse them I need to change some stuff which is not required for SVG, but it seems that it is not recommended to make a dependency to the image module to, if possible, to reuse the required panels. So is this correct and why?

@eirikbakke thx and no problem, I will change it as soon as possible :).

Chris2011 commented:
So I changed from XML first to image, then I moved from image to a new module called image.svg, now I called it just svg and the name is SVG Support as we have it for XML Support.

Next steps are:
* Add preview to navigator
* Add toolbar to the preview for scaling options.

FYI

Chris2011 commented:
@eirikbakke if I understand it correctly, there is no such a JSVGCanvas as we have it in batik so I need to create a BufferedImage as seen in this example from the readme: https://github.com/weisJ/jsvg and the quality is not that nice, because it will be converted to a pixel graphic.

I already added the settings:
g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
g.setRenderingHint(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_PURE);

It is okish, but not perfect. Let's see how it fits now. Watching SVG graphics they look like pixel graphics is kind a confusing. IMHO

Chris2011 commented:
Ok i see it myself, forget the last comment. Svg needs to be rendered as an image anyway so I can have a look what is the JSVGCanvas is doing and try to adapt it somehow.

neilcsmith-net commented:
@Chris2011 there appear to be at least two examples of SVG Swing components in the readme.  You almost certainly don't want to use `BufferedImage`.

eirikbakke commented:
@Chris2011 Yeah, I agree with @neilcsmith-net, normally you just paint the SVG directly in the component's paint() method; no need to go through a BufferedImage. That way you automatically get the right behavior on HiDPI screens, and don't have to allocate a buffer. You'd parse the SVG once to a com.github.weisj.jsvg.SVGDocument and then use that object to paint (rather than re-parsing on every paint).

Chris2011 commented:
Update here:
* Move everything to a new module
* Add toolbar with zoom, scaling (Same as for image, bit changes) and add custom background actions (similar to vs code) 
![image](https://github.com/user-attachments/assets/7763cb1b-abe8-42d8-8d2a-6376f8a8871a)
* Add thumbnail view into navigator

Chris2011 commented:
Due to the failing paperwork action I can do a clean PR with fork from current main and just add the correct changes w/o the update of batik and the other changes that I did before. What do you think? Then I also can use the correct name in the commit. NetBeans shows me another one from my work as default. So I will use the name that you know.

matthiasblaesing commented:
Nice! Some quick notes:

- Initialy the SVG document was not opened with the now support. If I see this correctly the module needs to be eager. With that change it works correctly:
```diff
--- a/ide/svg/nbproject/project.properties
+++ b/ide/svg/nbproject/project.properties
@@ -1,2 +1,3 @@
+is.eager=true
 javac.source=1.8
 javac.compilerargs=-Xlint -Xlint:-serial
```
- Source files missing a license header:
  - `BackgroundIcon.svg`
  - `SVGPanel.java`
  - `build.xml`
- The document icon `svgLogo.png` should have the svg sitting beside it, so that it can be rendered with SVG icon support in NetBeans
- The license info for `svgLogo.png` should use comment `COMMENT_UNSUPPORTED` (see https://github.com/apache/netbeans/blob/master/nbbuild/antsrc/org/netbeans/nbbuild/extlibs/licenseinfo/CommentType.java#L22-L25)
- Given this module is new, I would go to source level 17 directly
- Commit should hold full name as author (you can see the author here: https://patch-diff.githubusercontent.com/raw/apache/netbeans/pull/7938.patch or `git log -n 5` or with the History View from NetBeans
- there are remaining changes to dependencies in `idex/xml/nbproject/project.xml` 
- changes to `platform/libs.batik.read` can be reverted in a new commit. This should be squashed at the end anyway.
- Thumbnail support would be nice, but can be added later. The current state is already a very nice improvement. Thank you!
- Also nice-to-have: image viewer has middle mouse button drag, but can also added later.

Chris2011 commented:
> you don't _have_ to create a new PR since force pushing into PRs is OK. It could make it potentially more difficult for reviewers who use the github per-file "already viewed" feature for example - but you decide.

I may have expressed myself in a misleading way. I wanted to change this branch to -old and make a new with the same name to make this PR clean. But ok I see that I don't need to.

Thx @mbien and @matthiasblaesing will do changes that you all addressed.

@matthiasblaesing which thumbnail do you mean? I added a thumbnail functionality to show the SVG in navigator panel and this was/is working. But I see a bug that it is not updated when launching the dev IDE. But I can also have a look into this.

Funny that you said this with the middle mouse. I saw the code of dragging but it didn't work for me. Why? I have a trackpad and I just tried left and right click and dragging :D. I didn't try 3 fingers or middle mouse.

matthiasblaesing commented:
@Chris2011 sorry for the noise, navigator is empty directly after start. When I switched selected files, it was correctly updated.

eirikbakke commented:
> Should I load the *.svg or the *.png file? When I load *.svg in the dataObject and the multiViewElement, it is not working

SVG support in ImageUtilities, and NetBeans APIs that depend on it, works by automatically substituting ".svg" when you attempt to load a ".png" file. So make sure there is both a PNG file and an SVG file with the same base name, in the same folder (e.g. someFile.png and someFile.svg), and then refer to the PNG path rather than the SVG file path.

Chris2011 commented:
@eirikbakke thx for the info, I did this
![image](https://github.com/user-attachments/assets/44d7d6e2-23a4-4556-a2eb-57e9e38fd18a)

but it didn't show me the icon. I had this problem, when the name is the same but I will investigate deeper. Maybe a problem with the new implementation from me?

Chris2011 commented:
Just a comment, I also use the zoomIn/zoomOut icon from `org/netbeans/modules/image/zoomOut.gif`. Is this correct or should I copy them over to the new svg module?

eirikbakke commented:
> but it don't show me the icon. I have this problem, when the name is the same but I will investigate deeper. Maybe a problem with the new implementation from me?

Not sure; this seems like a layer registration problem rather than an SVG icon loading related problem. The icon shown in your screenshot is the general one for XML files, as you already noticed.

> Just a question, I also use the zoomIn/zoomOut icon from org/netbeans/modules/image/zoomOut.gif. Is this correct or should I copy them over to the new svg module?

I'm not actually sure what the rules are for when a module can access the resource files from another module. I'd suspect you can reference existing icons in modules that your own module has a dependency on, so long as those icons are in a public package of the referenced module.

Chris2011 commented:
> Not sure; this seems like a layer registration problem rather than an SVG icon loading related problem. The icon shown in your screenshot is the general one for XML files, as you already noticed.

No this is just the visual representation in my production IDE, this is the result in the IDE with the new Module
![image](https://github.com/user-attachments/assets/6f83d526-c3e1-49a5-86c9-12507d7afab8)

eirikbakke commented:
> No this was just the visual representation in my production IDE to show you that they have the same name.

Oh, interesting! Maybe try with just the PNG file present first, and then see if the presence of the SVG file causes the icon to disappear somehow (if the latter, we can debug further).

Chris2011 commented:
Just the png was working, my feeling was/is the same name but lemme check this first.

matthiasblaesing commented:
Ok, tested and works for me. I did this:

- check out this PR
- add `is.eager=true` to `project.properties` in `svg` module
- download the svg from https://github.com/apache/netbeans/pull/7938#issuecomment-2466238882 (icon provided by @eirikbakke )
- placed it in tree as `ide/svg/src/org/netbeans/modules/svg/resources/svgLogo.svg`: 
![grafik](https://github.com/user-attachments/assets/494cf384-e7fc-494e-b918-4c61303fe8b3)
- build from source
- opened the `svg` module in IDE started with `ant tryme` and got this: 
![grafik](https://github.com/user-attachments/assets/d33c67ce-7407-4448-8d86-f6d22fbe26d2)

To verify that the SVG is actually used, I ran with debug activated and placed a conditional break point in `platform/openide.util.ui/src/org/openide/util/ImageUtilities.java`, line 737, breaking on the condition `n.contains("svgLogo")`. That was matched and the loader code ine line 755 produced an icon.

Chris2011 commented:
@matthiasblaesing thx for testing and I will check this out.

Chris2011 commented:
Ok so I did an `ant clean` on my local netbeans repo and I see the svg. Maybe it was a caching problem. I think I'm ready in a couple of days. Thx for your help :)

matthiasblaesing reviewed (commented):
Made a second pass and looks really nice. I left a few inline comments and I noticed, that `svg` is missing from the `cluster.config`.

Chris2011 commented on ide/svg/licenseinfo.xml:
```diff
@@ -0,0 +1,31 @@
+-->
+<licenseinfo>
+  <fileset>
+    <file>src/org/netbeans/modules/svg/resources/svgLogo.svg</file>
```
Thx for the info I also see this in our codebase for SVG files, that's why I added the svg file too but no emotions, I can remove it. thx.

Chris2011 commented on ide/svg/licenseinfo.xml:
```diff
@@ -0,0 +1,31 @@
+    <file>src/org/netbeans/modules/svg/resources/svgLogo.svg</file>
+    <file>src/org/netbeans/modules/svg/resources/svgLogo.png</file>
+    <license ref="Apache-2.0-ASF" />
+    <comment type="COMMENT_UNSUPPORTED">
+      Images for the SVG filetype
+    </comment>
```
I saw this in the rust module, and I thought that could be necessary but you are right. WIll remove it.

Chris2011 commented on ide/svg/nbproject/project.properties:
```diff
@@ -0,0 +1,20 @@
+# under the License.
+
+is.eager=true
+javac.source=17
```
Oh, I changed this from the properties. Will add your change. thx

Chris2011 commented on ide/svg/src/org/netbeans/modules/svg/SVGDataObject.java:
```diff
@@ -0,0 +1,163 @@
+    displayName = "#LBL_Svg_LOADER",
+    position = 300
+)
+@ActionReferences(value = {
```
I will change this too, I already have some things todo so no problem here, I can change this. Thx.

Chris2011 commented:
Thx for the review, I also observed 2 things:
1. as @matthiasblaesing already mentioned, the navigation panel works after changing svg
2. the editor tab popup actions for just the preview has just two actions called editors and split. All others are available for source and history like Close, Close all, Close Others, etc.

If those should work for preview too and I would like to know, a hint would be nice. I tested it while pasting the actions registrations to the MultiViewElement but this didn't worked out.

matthiasblaesing commented:
For problem 1, this is the solution in `ide/image`: https://github.com/apache/netbeans/blob/9d5bc1e080daeac06ce0af3e6f1130003ef91015/ide/image/src/org/netbeans/modules/image/navigation/ImageNavigatorPanel.java#L137-L160

The problem is, that `setNewContent` and `getComponent` race. Normally the former wins and initial view does not happen, because you go out of the `null` check:

https://github.com/apache/netbeans/blob/2e720957f1554063363999ff5d938f1a0e556193/ide/svg/src/org/netbeans/modules/svg/navigation/SvgNavigatorPanel.java#L145-L147

I noticed this only after I removed my breakpoints out of the raced path and only into the "it is broken" regions.

For the actions: I noticed that too in other contexts. You might want to have a look at the implementation of the "History" multiview element, as that is in a similar role, than preview (not an editor, but should do many thing similar to the companion multiview element).

Chris2011 commented:
Thx again will check.

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SvgNavigatorPanel.java:
```diff
@@ -0,0 +1,209 @@
+
+                SwingUtilities.invokeLater(() -> panelUI.setSvg(svgDocument));
+            } catch (Exception ex) {
+                LOG.log(Level.SEVERE, ex.getMessage());
```
```java
                LOG.log(Level.SEVERE, "can not load svg: " + fo, ex);
```
retains stack and lists the file. Note that `SVGViewerElement` does also load SVGs, see other comment.

mbien replied:
try catch can be removed since loading does not throw exceptions, see https://github.com/apache/netbeans/pull/7938#discussion_r1893716745

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SvgNavigatorPanel.java:
```diff
@@ -0,0 +1,209 @@
+                        currentDataObject = DataObject.find(fe.getFile());
+                        setNewContent(currentDataObject);
+                    } catch (DataObjectNotFoundException ex) {
+                        Logger.getLogger(SvgNavigatorPanel.class.getName()).info(NbBundle.getMessage(SvgNavigatorPanel.class, "ERR_DataObject"));
```
can use outer class `LOG` constant

mbien commented on ide/svg/src/org/netbeans/modules/svg/toolbar/BackgroundIcon.java:
```diff
@@ -0,0 +1,77 @@
+        int width = getIconWidth();
+        int height = getIconHeight();
+
+        switch (bgMode) {
```
arrow switch candidate since exhaustive

mbien commented on ide/xml/nbproject/project.xml:
```diff
@@ -53,20 +53,20 @@
-                    <code-name-base>org.netbeans.modules.editor.lib</code-name-base>
+                    <code-name-base>org.netbeans.modules.editor.document</code-name-base>
                     <build-prerequisite/>
                     <compile-dependency/>
                     <run-dependency>
-                        <release-version>3</release-version>
-                        <specification-version>4.0</specification-version>
+                        <specification-version>1.0</specification-version>
                     </run-dependency>
```
nitpick: i believe the changes to the dependencies here are due to NB sorting the deps differently now, they can be likely reverted if all you wanted is to add the `friend-package` registration on the tail end of the file

Chris2011 replied:
Well I just added this via the UI. I can revert this.

mbien commented on ide/svg/src/org/netbeans/modules/svg/SVGViewerElement.java:
```diff
@@ -0,0 +1,351 @@
+            return;
+        }
+
+        svgDocument = svgLoader.load(fo.toURL());
```
not sure what the failure mode of SVGLoader is. Does it return null? Throw runtime exceptions? Javadoc unfortunately doesn't mention it. Probably something we should test by trying to load a broken svg file.

It would be better to extract loading into a utility method, which handles logging and Exceptions, so that both `SvgNavigatorPanel` and `SVGViewerElement` can use it.

mbien replied:
@weisJ thanks for the info! I haven't looked through it using an IDE until now so I wasn't sure about the details.

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SvgNavigatorPanel.java:
some classes use `SVG`, other use `Svg` prefix. I am ok with either but lets try to not mix styles within the same module at least.

mbien replied:
@Chris2011 you have to add the removed files to the commit otherwise the rename ends up being a copy operation.

Chris2011 replied:
I do everything as possible with NetBeans, which should do the work correctly, doesn't it? I thought NetBeans adds the stuff automatically when I commit.

mbien replied:
you can do that with NB sure. You also see the changeset below the commit window.

matthiasblaesing replied:
> I do everything as possible with NetBeans, which should do the work correctly, doesn't it? I thought NetBeans adds the stuff automatically when I commit.

It does, but sometimes the scanner is not quick enough to catch up. Either invoking refresh from a "Diff to" view or choosing a smaller set of files to commit helped in all cases I saw that happen.

mbien replied:
its also easy to overlook. E.g if the new files are added individually, the removed files won't be in the commit. Easiest way to avoid this is to commit via a parent folder. Right click on module/package and commit from there, this will include all changes. This is also how you can fix it if you forgot changes and amend the commit.

There are several other ways how to commit, e.g via show changes action and commit from there etc.

Chris2011 replied:
So hopefully I did it right this time.

Chris2011 commented:
> For the actions: I noticed that too in other contexts. You might want to have a look at the implementation of the "History" multiview element, as that is in a similar role, than preview (not an editor, but should do many thing similar to the companion multiview element).

I just had a look into the markdown support, because there it was working and it was just this code:

```
@Override
    public Action[] getActions() {
        return callback.createDefaultActions();
    }
```

and also the code to set callback. FYI.

matthiasblaesing commented:
I had a quick look at the current state and there are still duplicate classes: `SvgPreviewPanel` and `SVGPreviewPanel` and `SvgNavigatorPanel` and `SVGNavigatorPanel`.

Chris2011 commented:
The expected change is just renamed, right?

Chris2011 commented:
With the new commit, it seemed to work. I did this via git mv FYI. The last commit shows the renamed stuff and the files section just shows the correct file SVG*.java

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SVGNavigatorPanel.java:
```diff
@@ -0,0 +1,209 @@
+@NavigatorPanel.Registrations({
+    @NavigatorPanel.Registration(mimeType = "image/svg+xml", displayName = "#Navigator_DisplayName")
+})
+public class SvgNavigatorPanel implements NavigatorPanel {
```
does not match filename

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SVGPreviewPanel.java:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @author christian lenz
+ */
+public class SvgPreviewPanel extends JPanel {
```
does not match filename

Chris2011 replied:
Thx, my bad. I had no changes/uncommited stuff, that I checked and yes I need to check the build again which I didn't do this time. Sry for that.

mbien reviewed (commented):
please check that git shows no uncommitted changes in the area you worked on and that the module builds before pushing.

Chris2011 commented:
Fixed case problem, built, tested and everything works now.

matthiasblaesing reviewed (commented):
Thank you for the updates. I made a second pass and left a few inline comments. For these I build a fix here: https://github.com/apache/netbeans/commit/728260f6023abbaf6393f6d5df9c42bf09632548 (feel free to squash into your changes).

weisJ commented on ide/svg/src/org/netbeans/modules/svg/SVGPanel.java:
```diff
@@ -0,0 +1,129 @@
+            g2d.setRenderingHint(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_PURE);
+            g2d.scale(scale, scale);
+
+            svgDocument.render(this, g2d);
```
You might want to wrap this in a `try {} catch {}` block. Exceptions might happen due to bugs and you probably don't want the UI to crash in this case.

weisJ commented on ide/svg/src/org/netbeans/modules/svg/SVGPanel.java:
```diff
@@ -0,0 +1,129 @@
+
+        try {
+            g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+            g2d.setRenderingHint(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_PURE);
```
There are more SVG specific rendering hints in `com.github.weisj.jsvg.SVGRenderingHints`, which you might want to enable here. Specifically `KEY_MASK_CLIP_RENDERING` could probably be set to `VALUE_MASK_CLIP_RENDERING_ACCURACY` here. It will be a bit slower if masks are used, however only in this case it will produce the correct result in all case. The default is `VALUE_MASK_CLIP_RENDERING_FAST` which makes some assumptions about what features SVGs use to make some optimisations.

weisJ commented on ide/svg/src/org/netbeans/modules/svg/navigation/SVGPreviewPanel.java:
```diff
@@ -0,0 +1,95 @@
+            int scaledWidth = (int) (originalWidth / ratio);
+            int scaledHeight = (int) (originalHeight / ratio);
+
+            Graphics2D g2d = (Graphics2D) g.create((this.getWidth() - scaledWidth) / 2, (this.getHeight() - scaledHeight) / 2, scaledWidth, scaledHeight);
```
Rendering hints should also be set here to make rendering consistent across components.

Chris2011 commented on ide/svg/src/org/netbeans/modules/svg/toolbar/actions/CustomZoomAction.java:
```diff
@@ -0,0 +1,149 @@
+ */
+@ActionID(id = "org.netbeans.modules.svg.toolbar.CustomZoomAction", category = "View")
+@ActionRegistration(lazy = false, displayName = "#LBL_CustomZoom")
+public class CustomZoomAction extends CallableSystemAction {
```
Yes, I did this before as the image module did this too but I needed to call the corresponding actions for zooming. But I see how this is possible to do in another way. Thx for this new knowledge.

mbien commented on ide/svg/src/org/netbeans/modules/svg/SVGPanel.java:
```diff
@@ -0,0 +1,137 @@
+
+            g2d.scale(scale, scale);
+
+            svgDocument.render(this, g2d);
+        } catch (Exception ex) {
+            LOG.log(Level.SEVERE, ex.getMessage());
```
would be good if the panel would draw a text message to indicate that it can't render a svg.

I don't think we should log anything in an EDT draw method with level SEVERE since this is going to hit the notifications and will repeat. FINE or INFO is probably better here (FINE won't be logged by default - might be best? cc @matthiasblaesing ).

mbien commented on ide/svg/src/org/netbeans/modules/svg/SVGPanel.java:
```diff
@@ -0,0 +1,137 @@
+
+        Graphics2D g2d = (Graphics2D) g.create();
+
+        try {
+            g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+            g2d.setRenderingHint(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_PURE);
+            g2d.setRenderingHint(SVGRenderingHints.KEY_MASK_CLIP_RENDERING, SVGRenderingHints.VALUE_MASK_CLIP_RENDERING_ACCURACY);
+
+            g2d.scale(scale, scale);
```
move one line up so that is no longer in the try-catch

mbien commented on ide/svg/src/org/netbeans/modules/svg/SVGViewerElement.java:
```diff
@@ -0,0 +1,361 @@
+            return;
+        }
+
+        svgDocument = svgLoader.load(fo.toURL());
```
@weisJ will this load file-relative resources by default? if yes we should turn this off by default IMO.

was also mentioned in https://github.com/apache/netbeans/pull/7938#issuecomment-2466319810, cc @eirikbakke

weisJ replied:
No, loading of external resources needs to be enabled explicitly.

eirikbakke replied:
@weisJ Great, that's the desired behavior!
@mbien There's some extra logic in SVGIcon.java to deal with externally loaded resources, but it's only to force an exception to be thrown in these cases in order to make some existing unit tests pass. That's not needed in this PR.

Chris2011 commented:
@weisJ  While fixing the addressed feedback, I addressed an exception of course that this is rubbish because I took a file, renamed it to *.svg which was a *.http file with stuff that is not SVG. Shouldn't this be handled somehow? Yes I know, who the hell would do this. I wanted to test someting and this came up.

```
com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character '/' (code 47) in prolog; expected '<'
 at [row,col {unknown-source}]: [1,1]
	at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:666)
	at com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2148)
	at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1180)
	at com.ctc.wstx.evt.WstxEventReader.nextEvent(WstxEventReader.java:283)
[catch] at com.github.weisj.jsvg.parser.StaxSVGLoader.parse(StaxSVGLoader.java:106)
	at com.github.weisj.jsvg.parser.StaxSVGLoader.load(StaxSVGLoader.java:167)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:111)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:62)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:54)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:49)
	at org.netbeans.modules.svg.SVGViewerElement.updateView(SVGViewerElement.java:233)
	at org.netbeans.modules.svg.SVGViewerElement.componentOpened(SVGViewerElement.java:145)
	at org.netbeans.core.multiview.MultiViewPeer.showCurrentElement(MultiViewPeer.java:471)
	at org.netbeans.core.multiview.MultiViewPeer.showCurrentElement(MultiViewPeer.java:437)
	at org.netbeans.core.multiview.MultiViewPeer$SelectionListener.selectionChanged(MultiViewPeer.java:904)
	at org.netbeans.core.multiview.MultiViewModel.fireSelectionChanged(MultiViewModel.java:234)
	at org.netbeans.core.multiview.MultiViewModel.setActiveDescription(MultiViewModel.java:95)
	at org.netbeans.core.multiview.MultiViewModel$BtnGroup.setSelected(MultiViewModel.java:297)
	at java.desktop/javax.swing.JToggleButton$ToggleButtonModel.setSelected(JToggleButton.java:356)
	at java.desktop/javax.swing.ButtonGroup.setSelected(ButtonGroup.java:168)
	at org.netbeans.core.multiview.MultiViewModel$BtnGroup.setSelected(MultiViewModel.java:293)
	at org.netbeans.core.multiview.TabsComponent$ButtonMouseListener.mousePressed(TabsComponent.java:885)
	at java.desktop/java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:288)
	at java.desktop/java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:287)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6623)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3389)
	at java.desktop/java.awt.Component.processEvent(Component.java:6391)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5001)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4572)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:746)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:744)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:743)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

Chris2011 commented:
@weisJ

> Loading will never throw an exception. If parsing is not possible then null will be returned.

This is why I'm asking about the exceptions. I got a lot of exceptions is parsing is not possible. For example, exceptions will throw for content

```svg
<>
```

or without closing tag
``` svg
<svg>
```

or with this sample

```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg  version="1.1"
      baseProfile="full"
      width="300" height="200"
      xmlns="http://www.w3.org/2000/svg">

  <!-- Ein Rechteck in Rot -->
  <rect x="10" y="10" width="100" height="100" fill="red" />

  <!-- Hier der absichtlich fehlerhafte Kreis mit falsch geschlossenem Tag -->
  <g>
    <circle cx="150" cy="60" r="50" fill="blue"></circl>
  </g>
</svg>

```

weisJ commented:
I must have changed this since the last released version. That's my bad. In this case of course I would recommend to catch exceptions. I will try to release the next version soon.

Chris2011 commented:
Is this still possible to catch the error if the lib already does this? I already tried try/catch but I still get the exception. Is this done in another thread?

weisJ commented:
So is the exception thrown or simply logged? I am a bit confused about what the particular problem is here. Parsing is done on the same thread as `load` is called.

Chris2011 commented:
> So is the exception thrown or simply logged? I am a bit confused about what the particular problem is here. Parsing is done on the same thread as `load` is called.

This is what happens, it is shown in the notifications so I guess not just logged and I can't paint my error on that Panel due to the exception.

https://github.com/user-attachments/assets/378a0162-5f1b-4323-b6a0-4c07dff881be

mbien commented:
> I can't paint my error on that Panel due to the exception.

since you mentioned "painting the error",  you are probably looking at the wrong method. The trace shows that this is a svg **loading** exception, not an exception which happens during **painting**.

```
	at org.netbeans.modules.svg.navigation.SVGNavigatorPanel.lambda$setNewContent$1(SVGNavigatorPanel.java:149)
```

Chris2011 commented:
Yeah maybe it didn't show everysthing but we have more than one exception:

```
SEVERE [com.github.weisj.jsvg.parser.StaxSVGLoader]: Error while parsing SVG.
com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character '>' (code 62) in prolog, after '<'.
 at [row,col {unknown-source}]: [1,2]
	at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:666)
	at com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2177)
	at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1180)
	at com.ctc.wstx.evt.WstxEventReader.nextEvent(WstxEventReader.java:283)
[catch] at com.github.weisj.jsvg.parser.StaxSVGLoader.parse(StaxSVGLoader.java:106)
	at com.github.weisj.jsvg.parser.StaxSVGLoader.load(StaxSVGLoader.java:167)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:111)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:62)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:54)
	at com.github.weisj.jsvg.parser.SVGLoader.load(SVGLoader.java:49)
	at org.netbeans.modules.svg.SVGViewerElement.lambda$updateView$2(SVGViewerElement.java:238)
	at org.openide.util.RequestProcessor$Task.run(RequestProcessor.java:1403)
	at org.netbeans.modules.openide.util.GlobalLookup.execute(GlobalLookup.java:45)
	at org.openide.util.lookup.Lookups.executeWith(Lookups.java:287)
	at org.openide.util.RequestProcessor$Processor.run(RequestProcessor.java:2018)
```

This happens in the SVGViewerElement where I also run load in updateView.

mbien commented:
right. So usages of `load()` need to be called from a try/catch.

it might have been lost between so many messages but the timeline was like this if i remember correctly:
 - originally one of the calls had a try catch, i pointed that out and said something in the sense of "we have to check what the failure mode of a failing call is and might have to add it to both"
 - we got the info that the load call should not throw anything, but only return null, but it turned out that some exceptions slip through anyway
 - -> so lets add try / catch to the `load` usages again
 
 its off topic but IMO: anything what does IO and/or non-trivial parsing should require checked exceptions, this automatically documents how to handle failures etc, but its a minor detail here and we can't influence it anyway

Chris2011 commented:
> right. So usages of `load()` need to be called from a try/catch.
> ...
> * -> so lets add try / catch to the `load` usages again
> ...

I already do in my local code and still the error is thrown. This is why I'm wondering. In my code I already have try catch (locally). In the SVGViewerElementm in the SVGNavigatorPanel in the SVGPanel everywhere, I call load or render. But the screencapture is the result still.

mbien commented:
yes please share your changes somehow, if it doesn't build push into another branch and link it or something similar, otherwise simply add a commit to this PR like you did before.

Chris2011 commented:
> yes please share your changes somehow, if it doesn't build push into another branch and link it or something similar, otherwise simply add a commit to this PR like you did before.

It builds so no problem at all. I use the same branch so I'm still working on it. I also just add hints for me to fix this code. As you can see, I just wanted to show a JOptionPane for "debugging". This is my personal console.log :D.

mbien commented:
I can see how this can look confusing: jsvg itself logs with SEVERE level, this lands in the NB log handler.
https://github.com/search?q=repo%3AweisJ%2Fjsvg%20severe&type=code

All severe log lines (maybe also warnings - can't remember) will land in the notifications since they are automatically intercepted, as mentioned in https://github.com/apache/netbeans/pull/7938#discussion_r1898763135.

`load()` does in fact return null, no exception is actually caught by NB, we only see the log.

this experiment confirms it:
    
```java
    static{
        Logger.getLogger("com.github.weisj.jsvg.parser").setLevel(Level.OFF);
    }
```
mystery solved ;)

mbien commented on ide/svg/src/org/netbeans/modules/svg/navigation/SVGNavigatorPanel.java:
```diff
@@ -0,0 +1,208 @@
+                getComponent();
+            }
+
+            SVGLoader svgLoader = new SVGLoader();
+            SVGDocument svgDocument = svgLoader.load(fo.toURL());
+
+            if (svgDocument == null) {
+                return;
+            }
+
+            SwingUtilities.invokeLater(() -> panelUI.setSVG(svgDocument));
```
navigator needs to update even when svgDocument is null, otherwise it will keep displaying the last preview even if the selected file is broken - this could be confusing. The null check can be simply removed i think.

Chris2011 commented:
After my last changes I forgot to say that I have still one problem. It updates all messages on the opened SVG so filesChanged fired and everything works fine. When I switch to another SVG, resultChanged is fired but filesChanged never fired again so on save, navigator never updated. This is what I saw inside the debugger. So still working on it why this happens.

eirikbakke commented:
For your next revision, here is an updated version of the SVG icon, which matches the slightly darker background color I settled on for file icons in https://github.com/apache/netbeans/pull/8083.

![fileSVG](https://github.com/user-attachments/assets/7972c77e-85fd-40de-ba9c-1a8f30a2a69b) <-- here is the file

Chris2011 commented:
So first I just updated the svg, nothing more.
Second, the behavior that I mentioned, happens sometimes and I dunno whether this is just my dev env or real a bug. So pleaes can someone please confirm the problem that I have?

Update of navigator and svg preview is working after build and run the app. When the SVG was open before or a new one opens. When I have open multiple SVGs and I switch to another, just the svg preview updates, not the navigator.

Fun fact, when I click in the project tab around and go to the SVG which was not working correctly, it is working. So this is why I just saying it is sometimes not working correctly. Please verify, I will go deeper, but for now I couldn't find a solution, just that sometimes the listeners are not triggered anymore.

mbien commented on ide/svg/src/org/netbeans/modules/svg/SVGViewerElement.java:
```diff
@@ -0,0 +1,382 @@
+@Messages("LBL_SVGViewer=Preview")
+public class SVGViewerElement implements MultiViewElement {
+
+    static {
+        // JSVG loader/parser is logging with SEVERE level which would land in the NB exception reporter
+        Logger.getLogger("com.github.weisj.jsvg").setLevel(Level.OFF);
+    }
```
thought about this a bit more: it might be better to set this via the [config file](https://github.com/apache/netbeans/blob/daa8132538e9343d013cbac3c20cff859aefa5f4/nb/ide.launcher/netbeans.conf#L59). Since turning the logger on again might help @eirikbakke to debug icon loading.

I might be looking at the wrong places but I can't find any logging configs anywhere - a bit surprising that NB didn't have to adjust log levels of third party libs before. cc @neilcsmith-net

eirikbakke replied:
Perhaps set it to level Level.INFO instead? That way the error ends up on the console but does not trigger the exception reporter in the user interface.

mbien replied:
i think INFO is the default. Keep in mind that this setting is the cutoff for the log handler and does not influence the message log levels. The lib logs with SEVERE, unfortunately the next higher level is OFF.

eirikbakke replied:
> Keep in mind that this setting is the cutoff for the log handler and does not influence the message log levels. 

Oh... you're right. Idea: Turn the logging off only temporarily while the file is being loaded for SVG Preview purposes, but then turn it back to its previous value in a "finally" block right after.

mbien replied:
wouldn't recommend doing that for two reasons: threading and it being brittle since the places where the lib logs with SEVERE could change. Turning logging off for that package is just configuration - thats why I thought to move it to the NB config.
 
the spec is that it returns null if it fails - we have to live with it i guess. For debugging reasons we can switch it on again.

eirikbakke replied:
OK, I'm fine with turning logging off for JSVG then, by configuration or otherwise.

Perhaps eventually we could request from @weisJ that JSVG reduces the log level from SEVERE to INFO for problems that are caused by invalid SVG inputs, rather than true unexpected conditions?

Chris2011 commented:
@mbien @matthiasblaesing you can now recheck, i fixed all problems. I was kind a lost for the decision whether we would suppress the jsvg errors or not. It is also just one commit.

mbien reviewed (approved):
haven't re-tested it but changes look good to me, thanks!

sorry for the delay, wanted to solve the nb-javac problems first

Chris2011 commented:
Thx to all and no problem. I will wait a bit, until end of the week and if there are no other concerns, I will merge this.
